### PR TITLE
Grafana UI: Grid component in the unstable package

### LINKED ISSUE #73468: GrafanaUI: Create `Grid` component in `unstable` folder
Add a `Grid` component to grafana/ui in the `unstable` folder so it can be used locally to gather feedback on our layout components.

For inspiration on what properties are needed here, [this spreadsheet](https://docs.google.com/spreadsheets/d/1rZVJrs9x7WIyd3RSrK-Z6i_knXweeYuzZx3DDueKpHA/edit#gid=0) has examples of how other design systems implement this.

### PR DESCRIPTION
**Pull request by eledobleefe**
**What is this feature?**

We're building the `Grid` component inside the `unstable` package, so we can test it using it in real code and therefore get feedback about its props and usage.

**Why do we need this feature?**

To have a `Grid` component available.

**Who is this feature for?**

Everyone interested in using a `Grid` component

**Which issue(s) does this PR fix?**:

Fixes #73468 

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: public/app/features/connections/tabs/ConnectData/CardGrid/CardGrid.tsx [hunk 1]
```diff
@@ -38,6 +32,7 @@ const getStyles = (theme: GrafanaTheme2) => ({
         "Figure Description Actions"
         "Figure    Meta     Actions"
         "Figure     -       Actions"`,
+    marginBottom: '0px !important',
```
Comment (JoaoSilvaGrafana) [inline public/app/features/connections/tabs/ConnectData/CardGrid/CardGrid.tsx:35]: Why is this needed?

File: packages/grafana-ui/src/components/Grid/GridItem.tsx [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import { useStyles2 } from '../../themes';
+
+export interface GridItemProps {
+    children: NonNullable<React.ReactNode>;
+    /** If false will the grid item inherits the display of the parent. 
+    If true, the display with 'contents' as value. See info: https://developer.mozilla.org/en-US/docs/Web/CSS/display#display_contents*/
+    displayContents?: boolean;
+    /** Specifies the column where the grid item starts within the grid */
+    columnStart?: number;
+    /** Specifies the  column where the grid item starts.
```
Comment (JoaoSilvaGrafana) [inline packages/grafana-ui/src/components/Grid/GridItem.tsx:13]:
```suggestion
    /** Specifies the  column where the grid item ends. 
```

File: public/app/core/components/NavLandingPage/NavLandingPage.tsx [hunk 3]
```diff
@@ -15,43 +13,23 @@ interface Props {
...
-            <section className={styles.grid}>
+            <Grid display="grid" gap={3} templateColumns="repeat(auto-fill, minmax(300px,1fr))" autoRows="138px">
               {children?.map((child) => (
-                <NavLandingPageCard
-                  key={child.id}
-                  description={child.subTitle}
-                  text={child.text}
-                  url={child.url ?? ''}
-                />
+                <GridItem key={child.id} displayContents>
```
Comment (JoaoSilvaGrafana) [inline public/app/core/components/NavLandingPage/NavLandingPage.tsx:25]: I'm not sure I understand the `displayContents` prop and also I don't think we need the `GridItem` if that is the only prop that is set right? I think we can remove it from all these examples you created unless it needs to specify which row/column it is rendered at?

File: packages/grafana-ui/src/components/Grid/Grid.internal.story.tsx [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+import { Stack } from '../../unstable';
+import { withCenteredStory } from '../../utils/storybook/withCenteredStory';
+
+import { Grid } from './Grid';
+import mdx from './Grid.mdx';
+import { GridItem } from './GridItem';
+
+const gridItem = (index: number, color: string, width?: string, height?: string) => {
+  const itemOrder = index + 1;
+  return <div style={{ backgroundColor: color, width, height, fontSize: '12px' }}>Grid item nº{itemOrder}</div>;
```
[the remark this answers is not in the document]
Comment (eledobleefe) [reply packages/grafana-ui/src/components/Grid/Grid.internal.story.tsx:13]: I didn't realise it is not so common 😅 Changed 👌

File: packages/grafana-ui/src/components/Grid/Grid.internal.story.tsx [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+          templateColumns="repeat(2, 1fr)"
+          templateRows="repeat(3, 50px)"
+          alignItems="center"
+          gap={2}
+        >
+          {Array.from({ length: 6 }).map((_, i) => gridItem(i, colorList[i], undefined, '50%'))}
+        </Grid>
+      </span>
+      <span>
+        <p>Align items stretch - Grid structure reference</p>
```
[the remark this answers is not in the document]
Comment (eledobleefe) [reply packages/grafana-ui/src/components/Grid/Grid.internal.story.tsx:130]: Yes, that is the point. I'll change the text to 'Item number' 👌

File: public/app/features/connections/tabs/ConnectData/CardGrid/CardGrid.tsx (hunk 1 above)
Comment (eledobleefe) [reply public/app/features/connections/tabs/ConnectData/CardGrid/CardGrid.tsx:35]: `Card` applies a margin-bottom that makes the spacing inconsistent so I got rid of it while working on the `Grid` but forgot to delete this line. Thanks for telling me ;)

File: public/app/core/components/NavLandingPage/NavLandingPage.tsx (hunk 3 above)
Comment (eledobleefe) [reply public/app/core/components/NavLandingPage/NavLandingPage.tsx:25]: As I understand it, `display: contents` makes the children of the grid items be considered also grid and therefore apply the inheritance. I'll change the examples.

File: packages/grafana-ui/src/components/Layout/Grid/Grid.mdx [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+### When to use
+
+Use the Grid component when you want to create structured and organized layouts where content or elements need to be aligned in rows and columns for clarity and consistency.
+
+### When not to use
+
+Use the `Stack` component instead for these use cases:
+
+- **Simple layourts:** When you need to arrange elements in a linear format, either vertically or horizontally.
```
Comment (mikkancso) [inline packages/grafana-ui/src/components/Layout/Grid/Grid.mdx:21]: typo

```suggestion
- **Simple layouts:** When you need to arrange elements in a linear format, either vertically or horizontally.
```

**Review by mikkancso — COMMENTED** [review]
Changes under /connections look good. Thank you for doing this!! :raised_hands:

File: packages/grafana-ui/src/components/Layout/Grid/Grid.tsx [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+  columnGap,
+  rowGap,
+  templateColumns,
+  templateRows,
+  alignItems,
+  justifyItems,
+  autoFlow,
+  autoRows,
+}: GridProps) => {
+  const styles = useStyles2(
```
Comment (tskarhed) [inline packages/grafana-ui/src/components/Layout/Grid/Grid.tsx:48]: Maybe you can incorporate the changes Josh made to useStyles2 with additional arguments here?

File: packages/grafana-ui/src/components/Layout/Grid/GridItem.tsx [hunk 8]
```diff
@@ -0,0 +1,58 @@
+import { css } from '@emotion/css';
+import React, { useCallback } from 'react';
+
+import { useStyles2 } from '../../../themes';
+
+export interface GridItemProps {
+  children: NonNullable<React.ReactNode>;
+  /** If false will the grid item inherits the display of the parent. 
+    If true, the display with 'contents' as value. See info: https://developer.mozilla.org/en-US/docs/Web/CSS/display#display_contents*/
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/Layout/Grid/GridItem.tsx:9]:
```suggestion
  /** If set to 'false,' the grid item will inherit the parent's display property. If set to 'true,' the display property will be set to 'contents'. More info: https://developer.mozilla.org/en-US/docs/Web/CSS/display#display_contents */
```
I had a hard time understanding the original comment so I tried to rephrase it a bit, but not sure if I got it right.

File: packages/grafana-ui/src/components/Layout/Grid/GridItem.tsx [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+  displayContents?: boolean;
+  /** Specifies the column where the grid item starts within the grid */
+  columnStart?: number;
+  /** Specifies the  column where the grid item ends. 
+    If its value is span <number> the item spans across the provider number of columns*/
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/Layout/Grid/GridItem.tsx:14]:
```suggestion
  /** Specifies the column where the grid item ends.
    If its value is `span ${number}`, the item spans across the provided number of columns */
```
Using `<number>` here was throwing an error in console, since apparently the storybook was interpreting it as a `<number>` HTML tag, which doesn't exist.

File: packages/grafana-ui/src/components/Layout/Grid/GridItem.tsx [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+  columnEnd?: number | `span ${number}`;
+  /** Specifies the row where the grid item starts within the grid */
+  rowStart?: number;
+  /** Specifies the row where the grid item ends.*/
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/Layout/Grid/GridItem.tsx:18]:
```suggestion
  /** Specifies the row where the grid item ends */
```
Consistent punctuation.

File: packages/grafana-ui/src/components/Layout/Grid/GridItem.tsx [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+  rowEnd,
+}: GridItemProps) => {
+  const styles = useStyles2(
+    useCallback(
+      () => getGridItemStyles(displayContents, columnStart, columnEnd, rowStart, rowEnd),
+      [displayContents, columnStart, columnEnd, rowStart, rowEnd]
+    )
+  );
+
+  return React.createElement('div', { className: styles.gridItem }, children);
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/Layout/Grid/GridItem.tsx:37]:
```suggestion
  return <div className={styles.gridItem}>{children}</div>;
```

File: public/app/features/plugins/admin/components/PluginList.tsx [hunk 12]
```diff
@@ -17,28 +15,15 @@ interface Props {
...
   const { pathname } = useLocation();
   const pathName = config.appSubUrl + (pathname.endsWith('/') ? pathname.slice(0, -1) : pathname);
-
   return (
-    <div className={cx(styles.container, { [styles.list]: isList })} data-testid="plugin-list">
-      {plugins.map((plugin) => (
-        <PluginListItem key={plugin.id} plugin={plugin} pathName={pathName} displayMode={displayMode} />
-      ))}
+    <div data-testid="plugin-list">
+      <Grid display="grid" gap={3} templateColumns={isList ? '1fr' : 'repeat(auto-fill, minmax(288px, 1fr))'}>
```
Comment (Clarity-89) [inline public/app/features/plugins/admin/components/PluginList.tsx:22]: Should the Grid be able to accept this attributes instead (also `aria-label`, among others), so we don't need to introduce these kinds of wrappers?

File: packages/grafana-ui/src/components/Layout/Grid/Grid.tsx [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+  /** Specifies the gutters between columns */
+  columnGap?: ThemeSpacingTokens;
+  /** Specifies the gutters between rows */
+  rowGap?: ThemeSpacingTokens;
+  /** Defines the columns of the grid */
+  templateColumns?: string;
+  /** Defines the rows of the grid */
+  templateRows?: string;
+  /** Defines the horizontal alignment of the grid items in each cell */
+  justifyItems?: ItemsAlignment;
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/Layout/Grid/Grid.tsx:26]: Is there a reason we're not using `CSSProperties` interface instead of defining these custom types? 
```suggestion
  justifyItems?: CSSProperties['justifyItems'];
```

File: packages/grafana-ui/src/components/Layout/Grid/Grid.internal.story.tsx [hunk 14]
```diff
@@ -0,0 +1,356 @@
...
+  const colorOrder = index >= 6 ? index - 6 : index;
+  const itemOrder = index + 1;
+  const colorList = [
+    theme.visualization.getColorByName('red'),
+    theme.visualization.getColorByName('orange'),
+    theme.visualization.getColorByName('purple'),
+    theme.visualization.getColorByName('blue'),
+    theme.visualization.getColorByName('yellow'),
+    theme.visualization.getColorByName('green'),
+  ];
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/Layout/Grid/Grid.internal.story.tsx:24]:
```suggestion
const colorList = ['red', 'orange', 'purple', 'blue', 'yellow', 'green'].map(theme.visualization.getColorByName);
```
Nit: This can be simplified a bit.

**Review by Clarity-89 — COMMENTED** [review]
Nice work on this! Left a few comments, mostly wonder if these components should extend the basic HTML attributes in their types, allowing to pass native HTML attributes as props.

File: public/app/features/plugins/admin/components/PluginList.tsx (hunk 12 above)
Comment (eledobleefe) [reply public/app/features/plugins/admin/components/PluginList.tsx:22]: @JoaoSilvaGrafana and I talked about this and agreed on not accepting them at this moment. We'll let devs use this component and give feedback, including if they need these attributes in the Grid. Correct me if I'm wrong @JoaoSilvaGrafana.

File: packages/grafana-ui/src/components/Layout/Grid/Grid.mdx [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+
+Use the `Stack` component instead for these use cases:
+
+- **Simple layouts:** When you need to arrange elements in a linear format, either vertically or horizontally.
+- **Dynamic content:** When your content or elements have varying heights or widths and you want them to adapt to each item's size without predefined rows or columns.
+
+Use the `Flex` component instead for these use cases:
+
+- **Centering:** Perfect for centering elements both vertically and horizontally.
+- **Dynamic Order:** Easily reorder elements for responsive layouts without changing code.
```
[the remark this answers is not in the document]
Comment (eledobleefe) [reply packages/grafana-ui/src/components/Layout/Grid/Grid.mdx:27]: To be honest, I copied the `Flex` part from the Stack component documentation.

File: packages/grafana-ui/src/components/Layout/Grid/Grid.internal.story.tsx [hunk 16]
```diff
@@ -0,0 +1,342 @@
...
+  gap: 2,
+};
+
+Basic.parameters = {
+  controls: {
+    exclude: ['display', 'alignItems', 'justifyItems', 'autoRows', 'columnStart', 'columnEnd', 'rowStart', 'rowEnd'],
+  },
+};
+
+export const AlignItemsExamples: StoryFn<typeof Grid> = (args) => {
```
Comment (L-M-K-B) [inline packages/grafana-ui/src/components/Layout/Grid/Grid.internal.story.tsx:94]: I understand that it is meant to show different examples for alignment but it looks off as soon as the width is a little smaller.

Screenshot 2023-09-19 at 16 58 44

File: packages/grafana-ui/src/components/Layout/Grid/Grid.internal.story.tsx [hunk 17]
```diff
@@ -0,0 +1,360 @@
...
+      },
+    },
+  },
+  args: {
+    display: 'grid',
+    gap: 1,
+  },
+};
+
+export const Basic: StoryFn<typeof Grid> = (args) => {
```
Comment (L-M-K-B) [inline packages/grafana-ui/src/components/Layout/Grid/Grid.internal.story.tsx:67]: Some options of the autoFlow prop look the same (row = row dense, column = column dense). Can you please check whether it's possible to make clear what the differences are?

File: packages/grafana-ui/src/components/Layout/Grid/Grid.mdx [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+import { Grid } from './Grid';
+import { GridItem } from './GridItem';
+
+<Meta title="MDX|Grid" component={Grid} />
+
+# Grid
+
+The `Grid` component allows for the organized layout and alignment of content into a grid-based structure.
+
+## Usage
```
Comment (L-M-K-B) [inline packages/grafana-ui/src/components/Layout/Grid/Grid.mdx:11]: In order to have a complete docs page I would expect that GridItem and possible use cases are mentioned here.

File: public/app/features/connections/tabs/ConnectData/CardGrid/CardGrid.tsx [hunk 19]
```diff
@@ -64,7 +58,7 @@ export const CardGrid = ({ items, onClickItem }: CardGridProps) => {
   const styles = useStyles2(getStyles);
 
   return (
-    <ul className={styles.sourcesList}>
+    <Grid display="grid" templateColumns="repeat(auto-fill, minmax(380px, 1fr))" gap={1.5}>
```
Comment (L-M-K-B) [inline public/app/features/connections/tabs/ConnectData/CardGrid/CardGrid.tsx:61]: Comparing the view in ops and locally it seems that there is a lot of unnecessary empty space now because the cards are arranged in only 2 columns until the breakpoint of approx. 1500px (before there were three, breakpoint from 2 to 3 columns 1250px). Is this intentional?

View on localhost:
Screenshot 2023-09-19 at 17 40 57

View in ops:
Screenshot 2023-09-19 at 17 41 46

File: packages/grafana-ui/src/components/Layout/Grid/GridItem.tsx [hunk 20]
```diff
@@ -0,0 +1,57 @@
+import { css } from '@emotion/css';
+import React, { useCallback } from 'react';
+
+import { useStyles2 } from '../../../themes';
+
+export interface GridItemProps {
```
Comment (L-M-K-B) [inline packages/grafana-ui/src/components/Layout/Grid/GridItem.tsx:6]: There are no props for further styling yet (e.g. background colour, font size or colour, etc.) which forces users of GridItem to create another div inside. Is this something we can add or is the intention here to keep the component as basic as possible?

**Review by L-M-K-B — COMMENTED** [review]
This looks great already! However I added a couple of comments and questions which arose during review.

File: packages/grafana-ui/src/components/Layout/Grid/Grid.internal.story.tsx (hunk 16 above)
Comment (eledobleefe) [reply packages/grafana-ui/src/components/Layout/Grid/Grid.internal.story.tsx:94]: Yes, I was thinking about changing the name to `Align items stretch (reference)`. Will it be understandable? 🤔

File: packages/grafana-ui/src/components/Layout/Grid/Grid.internal.story.tsx (hunk 17 above)
Comment (eledobleefe) [reply packages/grafana-ui/src/components/Layout/Grid/Grid.internal.story.tsx:67]: It seems it reorders the items to try to fill the holes first so the smaller items go last. It is explained [here](https://css-tricks.com/snippets/css/complete-guide-grid/#aa-grid-auto-flow). Will it be useful to show this as an example? This is a normal property of CSSGrid 🤷

File: packages/grafana-ui/src/components/Layout/Grid/GridItem.tsx (hunk 20 above)
Comment (eledobleefe) [reply packages/grafana-ui/src/components/Layout/Grid/GridItem.tsx:6]: We want to keep it as basic as possible. Also, when using GridItem we are setting `display: contents`. It behaves like it does not occupy any space so the element to have the styling is the child itself.

File: public/app/features/connections/tabs/ConnectData/CardGrid/CardGrid.tsx (hunk 19 above)
Comment (L-M-K-B) [reply public/app/features/connections/tabs/ConnectData/CardGrid/CardGrid.tsx:61]: This seems to be a general issue. I talked to @teddyba as well as @besartberisha and I created an issue for the responsible team: https://github.com/grafana/cloud-onboarding/issues/5013

**Comment by eledobleefe** [comment]
Closing this pull request in favour of this one #75299.
